File: 3.x.x/crates-utopia/user-agent/src/user_agent.rs
use std::sync::OnceLock;

use serde::Serialize;

use crate::bot::Bot;
use crate::client::Client;
use crate::detection;
use crate::device::Device;
use crate::operating_system::OperatingSystem;

/// A lazily evaluated analysis of one user-agent string.
///
/// Each category is detected at most once. Bot detection is independent from
/// client and device detection, so bots never suppress the other results.
#[derive(Debug)]
pub struct UserAgent {
    value: String,
    operating_system: OnceLock<OperatingSystem>,
    client: OnceLock<Client>,
    device: OnceLock<Device>,
    bot: OnceLock<Option<Bot>>,
}

/// Nested `UserAgent::to_array()` output.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct UserAgentArray {
    pub os: crate::operating_system::OperatingSystemArray,
    pub client: crate::client::ClientArray,
    pub device: crate::device::DeviceArray,
    pub bot: Option<crate::bot::BotArray>,
}

impl UserAgent {
    /// Parse a user-agent string for lazy detection.
    pub fn parse(value: &str) -> Self {
        Self {
            value: value.to_string(),
            operating_system: OnceLock::new(),
            client: OnceLock::new(),
            device: OnceLock::new(),
            bot: OnceLock::new(),
        }
    }

    /// Original user-agent string.
    pub fn raw(&self) -> &str {
        &self.value
    }

    /// Detected operating system (memoized).
    pub fn operating_system(&self) -> OperatingSystem {
        self.operating_system
            .get_or_init(|| detection::detect_operating_system(&self.value))
            .clone()
    }

    /// Detected client (memoized).
    pub fn client(&self) -> Client {
        self.client
            .get_or_init(|| detection::detect_client(&self.value))
            .clone()
    }

    /// Detected device (memoized).
    pub fn device(&self) -> Device {
        self.device
            .get_or_init(|| detection::detect_device(&self.value))
            .clone()
    }

    /// Detected bot, if any (memoized).
    pub fn bot(&self) -> Option<Bot> {
        self.bot
            .get_or_init(|| detection::detect_bot(&self.value))
            .clone()
    }

    /// Whether the user-agent matches a known bot.
    pub fn is_bot(&self) -> bool {
        self.bot().is_some()
    }

    /// Nested serialization matching PHP `toArray()`.
    pub fn to_array(&self) -> UserAgentArray {
        UserAgentArray {
            os: self.operating_system().to_array(),
            client: self.client().to_array(),
            device: self.device().to_array(),
            bot: self.bot().map(|b| b.to_array()),
        }
    }
}

File: packages/user-agent/src/UserAgent/UserAgent.php
<?php

declare(strict_types=1);

namespace Utopia\UserAgent;

use Utopia\UserAgent\Detection\BotDetector;
use Utopia\UserAgent\Detection\ClientDetector;
use Utopia\UserAgent\Detection\DeviceDetector;
use Utopia\UserAgent\Detection\OperatingSystemDetector;

/**
 * A lazily evaluated analysis of one user-agent string.
 *
 * Each category is detected at most once. Bot detection is independent from
 * client and device detection, so bots never suppress the other results.
 */
final class UserAgent
{
    private ?OperatingSystem $operatingSystem = null;

    private ?Client $client = null;

    private ?Device $device = null;

    private ?Bot $bot = null;

    private bool $botResolved = false;

    private function __construct(private readonly string $value) {}

    public static function parse(string $value): self
    {
        return new self($value);
    }

    public function raw(): string
    {
        return $this->value;
    }

    public function operatingSystem(): OperatingSystem
    {
        return $this->operatingSystem ??= OperatingSystemDetector::detect($this->value);
    }

    public function client(): Client
    {
        return $this->client ??= ClientDetector::detect($this->value);
    }

    public function device(): Device
    {
        return $this->device ??= DeviceDetector::detect($this->value);
    }

    public function bot(): ?Bot
    {
        if (!$this->botResolved) {
            $this->bot = BotDetector::detect($this->value);
            $this->botResolved = true;
        }

        return $this->bot;
    }

    public function isBot(): bool
    {
        return $this->bot() instanceof \Utopia\UserAgent\Bot;
    }

    /**
     * @return array{
     *     os: array{code: ?string, name: ?string, version: ?string},
     *     client: array{type: ?string, code: ?string, name: ?string, version: ?string, engine: ?string, engineVersion: ?string},
     *     device: array{type: ?string, brand: ?string, model: ?string},
     *     bot: array{name: string, category: string}|null
     * }
     */
    public function toArray(): array
    {
        return [
            'os' => $this->operatingSystem()->toArray(),
            'client' => $this->client()->toArray(),
            'device' => $this->device()->toArray(),
            'bot' => $this->bot()?->toArray(),
        ];
    }
}
